### ISSUE TITLE
How to debug cpp files? breakpoints not working

### ISSUE THREAD

**Issue by mehran22k** [BelledonneCommunications/linphone-android#674]
Hi

I need to watch some variable in **linphone-sdk-android/linphone-sdk/linphone/src/db/main-db.cpp**  in debug
How can i do it?
I set some breakpoints but they re not fired

I build SDK locally
And also done [this ](https://github.com/BelledonneCommunications/linphone-android#native-debugging)instruction but it s also not working for me 

Please help me
Thank you in advance

**Comment by Viish** [comment]
Hi,
Actually https://github.com/BelledonneCommunications/linphone-android#native-debugging is the correct way to debug native libraries in Android studio.
Make sure your Android Studio is up to date, and maybe try with the emulator instead of a device or vice-versa.
If you find an issue in our native debugging protocol, feel free to share it, thanks.
Cheers,

**Comment by Viish** [comment]
You can also try https://wiki.linphone.org/xwiki/wiki/public/view/Lib/Getting%20started/Android/#HNativedebugging but I think it's outdated now.

**Comment by mehran22k** [comment]
Hi @Viish 
Thank you so much

For example, If i want to debug cpp files in **linphone-android\linphone-sdk-android\linphone-sdk\linphone\src\chat\chat-message**


Then , ( **"Select 'Dual' or 'Native' and add the path to linphone-sdk libraries."**) I have to add a path. so I add this path:
linphone-android\linphone-sdk-android\linphone-sdk\linphone\src\
but its not working.
Am i right?

**Comment by Viish** [comment]
No.
And the path to the .so directory that matches your abi.
In your case: linphone-android\linphone-sdk-android\linphone-sdk\build\libs\<abi>

**Comment by mehran22k** [comment]
Thank you for your great help
I add this path and its worker now:
linphone-android\linphone-sdk-android\linphone-sdk\build\**libs-debug**<abi>

**Comment by sitio-couto** [comment]
Hey @Viish ,

I'm having this exact same problem, but the path to the debug symbols seems to be correct.

When setting the debug mode to **dual**, breakpoints in C++ code are not functional.

When setting the debug mode to **native only**, breakpoints in C++ show the following warning:

```
~/linphone-sdk/liblinphone/src/factory/factory.cpp:552  The breakpoint will not currently be hit. No executable code is associated with this line.
```

I followed the README section for both building a local android SDK as well as for native debugging:

1. Build the Android SDK using the container method with the following CMake configurations:
```
cmake .. -DCMAKE_BUILD_TYPE=Debug  -DLINPHONESDK_PLATFORM=Android -DLINPHONESDK_ANDROID_ARCHS=x86
```
1. Set the debugger to Dual.
1. Add `~/linphone-sdk/build/libs-debug` as a _Symbol Directory_.
1. Open `~/linphone-sdk/liblinphone/src/factory/factory.cpp` and set a breakpoint at `Factory::createConfigWithFactory` (line 552).
1. Run the app in debug mode using a Pixel 5 x86 AVD as the target.

The breakpoint, however, is never reached.

To ensure the code with the breakpoint is being executed, I've replaced the return in line 552 with `return nullptr;` in the local SDK and rebuilt it. The app crashed with a null pointer exception, as expected.

Any ideas of what I might be missing?
